File: src/application/models/Apprentice.java
package application.models;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import application.ROM;
import application.controllers.MainMenu;
import application.models.lists.Characters;

public class Apprentice extends Model implements Serializable
{
    private List<EquipmentAtLevel>  equipmentAtLevels          = new ArrayList<EquipmentAtLevel>();
    private int[][]		            dialogues;
    private static List<Apprentice> models;
    private static List<Apprentice> Dmodels;
    private static final int	    apprenticeListSize         = 7;
    private int			            equipCode;
    private int[]		            dialoguePointers;
    private static final int	    baseOffset		           = 0x8740;
    private static final int	    bytesPerApprentice         = 60;
    private static final int	    baseEquipmentOffset	       = 0x88E4;
    private static final int	    bytesPerEquipmentList      = 44;
    private static final int	    equipmentRowsPerApprentice = 10;
    private static final int	    numOfDialogues	       = 6;

    public Apprentice(int gameIndex)
    {
	super(gameIndex);
	name = Characters.getCharacterNames()[gameIndex];
	equipCode = Characters.getEquipCodes()[gameIndex];
    }

    public void getValuesFromROM()
    {
	dialoguePointers = new int[numOfDialogues];
	ROM.setOffset(baseOffset + bytesPerApprentice * gameIndex);
	for(int i=0; i<numOfDialogues;i++)
	{
		dialoguePointers[i] = ROM.getNextTriple();
	}
	name = Characters.getCharacterNames()[gameIndex];

	dialogues = new int[apprenticeListSize][numOfDialogues];

	for (int i = 0; i < apprenticeListSize; i++)
	{
	    for (int k = 0; k < numOfDialogues; k++)
	    {
		dialogues[i][k] = ROM.getNextByte();
	    }
	}

	ROM.setOffset(baseEquipmentOffset + bytesPerEquipmentList * gameIndex);
	for (int i = 0; i < equipmentRowsPerApprentice; i++)
	{
	    equipmentAtLevels.add(new EquipmentAtLevel(ROM.getNextByte(), new int[]
	    {
				ROM.getNextByte(), ROM.getNextByte(), ROM.getNextByte()
	    }));
	}
    }

    public void writeValuesToROM()
    {
	ROM.setOffset(baseOffset + 18 + bytesPerApprentice * gameIndex);

	for (int i = 0; i < apprenticeListSize; i++)
	{
	    for (int k = 0; k < numOfDialogues; k++)
	    {
		ROM.setNextByte(dialogues[i][k]);
	    }
	}

	ROM.setOffset(baseEquipmentOffset + bytesPerEquipmentList * gameIndex);

	for (int i = 0; i < equipmentRowsPerApprentice; i++)
	{
	    for (EquipmentAtLevel entry : equipmentAtLevels)
	    {
		ROM.setNextByte(entry.getLevel());
		ROM.setNextByte(entry.getEquipment()[0]);
		ROM.setNextByte(entry.getEquipment()[1]);
		ROM.setNextByte(entry.getEquipment()[2]);
	    }
	}
    }

    public List<EquipmentAtLevel> getEquipmentAtLevels()
    {
	return equipmentAtLevels;
    }

    public void setEquipmentAtLevels(List<EquipmentAtLevel> equipmentAtLevels)
    {
	this.equipmentAtLevels = equipmentAtLevels;
    }

    public int[][] getDialogues()
    {
	return dialogues;
    }

    public static List<Apprentice> getModels()
    {
	return models;
    }

    public static List<Apprentice> getDModels()
    {
	return Dmodels;
    }

    public static void setDModels(List<Apprentice> defaultApprentices)
    {
	Apprentice.Dmodels = defaultApprentices;
    }

    public int getEquipCode()
    {
	return equipCode;
    }

    public int[] getDialoguePointers()
    {
	return dialoguePointers;
    }

}

File: src/application/models/Apprentice.kt
package application.models

import application.ROM
import application.models.lists.Characters
import java.io.Serializable
import java.util.*

class Apprentice(gameIndex: Int) : Model(gameIndex), Serializable
{
	var equipmentAtLevels: MutableList<EquipmentAtLevel> = ArrayList()
	var dialogues = Array(apprenticeListSize) { IntArray(numOfDialogues) }
	val equipCode = Characters.equipCodes[gameIndex]
	var dialoguePointers = IntArray(numOfDialogues)

	init
	{
		name = Characters.characterNamesByGameIndex[gameIndex]
	}

	fun getValuesFromROM()
	{

		ROM.offset = baseOffset + bytesPerApprentice * gameIndex
		offset = ROM.offset
		for (i in 0..numOfDialogues - 1)
		{
			dialoguePointers[i] = ROM.nextTriple
		}
		name = Characters.characterNamesByGameIndex[gameIndex]



		for (i in 0..apprenticeListSize - 1)
		{
			for (k in 0..numOfDialogues - 1)
			{
				dialogues[i][k] = ROM.nextByte
			}
		}

		ROM.offset = baseEquipmentOffset + bytesPerEquipmentList * gameIndex
		for (i in 0..equipmentRowsPerApprentice - 1)
		{
			equipmentAtLevels.add(EquipmentAtLevel(ROM.nextByte, intArrayOf(ROM.nextByte, ROM.nextByte, ROM.nextByte)))
		}
	}

	fun writeValuesToROM()
	{
		ROM.offset = baseOffset + 18 + bytesPerApprentice * gameIndex

		for (i in 0..apprenticeListSize - 1)
		{
			for (k in 0..numOfDialogues - 1)
			{
				ROM.nextByte = dialogues[i][k]
			}
		}

		ROM.offset = baseEquipmentOffset + bytesPerEquipmentList * gameIndex

		for (i in 0..equipmentRowsPerApprentice - 1)
		{
			for (entry in equipmentAtLevels)
			{
				ROM.nextByte = entry.level
				ROM.nextByte = entry.equipment[0]
				ROM.nextByte = entry.equipment[1]
				ROM.nextByte = entry.equipment[2]
			}
		}
	}


	companion object
	{
		private val apprenticeListSize = 7
		private val baseOffset = 0x8740
		private val bytesPerApprentice = 60
		private val baseEquipmentOffset = 0x88E4
		private val bytesPerEquipmentList = 44
		private val equipmentRowsPerApprentice = 10
		private val numOfDialogues = 6
	}

}
